In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings

warnings.filterwarnings('ignore')
sns.set_style('whitegrid')

import phik

In [2]:
import json

In [3]:
import datetime as DT

In [4]:
from sklearn.metrics import f1_score, accuracy_score, precision_score
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.metrics import make_scorer, precision_score, r2_score
from sklearn.model_selection import train_test_split, GridSearchCV, TimeSeriesSplit
from sklearn.preprocessing import StandardScaler, TargetEncoder

In [5]:
data = pd.read_csv('data1.csv')
data

date                             st_id  \
0        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
1        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
2        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
3        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
4        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
...             ...                               ...   
2124667  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124668  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124669  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124670  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124671  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   

                                pr_sku_id  pr_sales_type_id  price  \
0        0045ebdb1069ff4b3dd3efe628c39cd3                 0    0.0   
1        005c52de11abaf8e0a9714b24415ce34                 0    0.0   
2        00661699f543753ec7e911a64b9fd2f6                 0  435.0   
3        0094042bfeae507dc7f62acc8e5ed03a                 1  191.5   
4        0120fa9d93da50a73a6f61838b49ff46                 0    0.0   
...                                   ...               ...    ...   
2124667  fe50ae64d08d4f8245aaabc55d1baf79                 1  143.2   
2124668  fe5d18ae6650335830e4c1dbd9e6ddb9                 0    0.0   
2124669  ff1f1e5d2708809c03d03ff8b689030b                 0    0.0   
2124670  ff5cb535afe362f766ca6bd5d8e92e88                 0   46.0   
2124671  ff62e7bffaca5e0ac8029cd29298c8c3                 0    0.0   

         total_sales                        st_city_id  \
0                0.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
1                0.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
2                1.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
3                6.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
4                0.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
...              ...                               ...   
2124667          5.0  885fe656777008c335ac96072a45be15   
2124668          0.0  885fe656777008c335ac96072a45be15   
2124669          0.0  885fe656777008c335ac96072a45be15   
2124670          3.0  885fe656777008c335ac96072a45be15   
2124671          0.0  885fe656777008c335ac96072a45be15   

                         st_division_code  st_type_format_id  st_type_loc_id  \
0        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
1        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
2        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
3        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
4        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
...                                   ...                ...             ...   
2124667  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124668  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124669  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124670  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124671  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   

         st_type_size_id  st_is_active  sales_total  \
0                      8             1     143898.0   
1                      8             1     143898.0   
2                      8             1     143898.0   
3                      8             1     143898.0   
4                      8             1     143898.0   
...                  ...           ...          ...   
2124667               12             1     100699.0   
2124668               12             1     100699.0   
2124669               12             1     100699.0   
2124670               12             1     100699.0   
2124671               12             1     100699.0   

                              pr_group_id                         pr_cat_id  \
0        c74d97b01eae257e44aa9d5bade97baf  c559da2ba967eb820766939a658022c8   
1        c20

In [6]:

with open('segment2cluster_43.json', 'r') as json_file:
    clusters = json.load(json_file)

In [7]:
print(clusters)

{'16a5cdae362b8d27a1d8f8c7b78b4330_0045ebdb1069ff4b3dd3efe628c39cd3': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_005c52de11abaf8e0a9714b24415ce34': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_00661699f543753ec7e911a64b9fd2f6': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_0094042bfeae507dc7f62acc8e5ed03a': 11, '16a5cdae362b8d27a1d8f8c7b78b4330_0120fa9d93da50a73a6f61838b49ff46': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_0169529ff660adcac9b7e354e0c4b882': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_01e4734745e97e52d3213449e1a05dd7': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_02713435f3587e2c81d8f6a9016763ea': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_028601bcf68af10f8be7422c22b5f6a6': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_033013f94a18c066e8b3d610bed34bee': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_040a02c2ad1561cbcfc9cae5b4c4b73b': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_0448c8afc036bb44c457d7e9edd74c50': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_04bbb07b1057b09d04209991f3eadd8f': 9, '16a5cdae362b8d27a1d8f8c7b78b4330_050a8120e2a3b8751df8752fcc03

In [8]:
for index, row in data.iterrows():
    segment = row["st_id"] + "_" + row["pr_sku_id"]
    if segment in clusters:
        data.at[index, 'cluster'] = clusters[segment]
    else:
        data.at[index, 'cluster'] = None  # Or any default value if the segment is not found in the dictionary


In [9]:
display(data)

date                             st_id  \
0        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
1        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
2        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
3        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
4        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
...             ...                               ...   
2124667  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124668  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124669  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124670  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124671  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   

                                pr_sku_id  pr_sales_type_id  price  \
0        0045ebdb1069ff4b3dd3efe628c39cd3                 0    0.0   
1        005c52de11abaf8e0a9714b24415ce34                 0    0.0   
2        00661699f543753ec7e911a64b9fd2f6                 0  435.0   
3        0094042bfeae507dc7f62acc8e5ed03a                 1  191.5   
4        0120fa9d93da50a73a6f61838b49ff46                 0    0.0   
...                                   ...               ...    ...   
2124667  fe50ae64d08d4f8245aaabc55d1baf79                 1  143.2   
2124668  fe5d18ae6650335830e4c1dbd9e6ddb9                 0    0.0   
2124669  ff1f1e5d2708809c03d03ff8b689030b                 0    0.0   
2124670  ff5cb535afe362f766ca6bd5d8e92e88                 0   46.0   
2124671  ff62e7bffaca5e0ac8029cd29298c8c3                 0    0.0   

         total_sales                        st_city_id  \
0                0.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
1                0.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
2                1.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
3                6.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
4                0.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
...              ...                               ...   
2124667          5.0  885fe656777008c335ac96072a45be15   
2124668          0.0  885fe656777008c335ac96072a45be15   
2124669          0.0  885fe656777008c335ac96072a45be15   
2124670          3.0  885fe656777008c335ac96072a45be15   
2124671          0.0  885fe656777008c335ac96072a45be15   

                         st_division_code  st_type_format_id  st_type_loc_id  \
0        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
1        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
2        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
3        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
4        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
...                                   ...                ...             ...   
2124667  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124668  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124669  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124670  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124671  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   

         st_type_size_id  st_is_active  sales_total  \
0                      8             1     143898.0   
1                      8             1     143898.0   
2                      8             1     143898.0   
3                      8             1     143898.0   
4                      8             1     143898.0   
...                  ...           ...          ...   
2124667               12             1     100699.0   
2124668               12             1     100699.0   
2124669               12             1     100699.0   
2124670               12             1     100699.0   
2124671               12             1     100699.0   

                              pr_group_id                         pr_cat_id  \
0        c74d97b01eae257e44aa9d5bade97baf  c559da2ba967eb820766939a658022c8   
1        c20

In [10]:
data['cluster'].isna().value_counts()
data['cluster'].value_counts()

9.0     1958528
11.0      74624
0.0       23232
7.0       19712
34.0       9856
4.0        7392
5.0        4224
8.0        3168
3.0        2816
1.0        2112
13.0       1760
18.0       1408
17.0       1408
15.0       1056
10.0       1056
16.0       1056
2.0        1056
14.0        704
6.0         704
12.0        704
27.0        352
21.0        352
19.0        352
38.0        352
22.0        352
31.0        352
33.0        352
32.0        352
36.0        352
37.0        352
40.0        352
25.0        352
29.0        352
28.0        352
42.0        352
41.0        352
39.0        352
30.0        352
24.0        352
35.0        352
26.0        352
23.0        352
20.0        352
Name: cluster, dtype: int64

In [11]:
check = data[data['cluster']==11.0].copy()
check

date                             st_id  \
3        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
13       2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
34       2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
70       2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
85       2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
...             ...                               ...   
2124477  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124488  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124558  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124597  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124667  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   

                                pr_sku_id  pr_sales_type_id  price  \
3        0094042bfeae507dc7f62acc8e5ed03a                 1  191.5   
13       050a8120e2a3b8751df8752fcc0323ef                 1  162.3   
34       0cf7b480289c0c4b07b6d3bd72fef0c9                 0    0.0   
70       194adbcaa46104c43d41ad4f23c18407                 1   53.0   
85       1cec786e41a51f75b5701511f995218b                 1   67.2   
...                                   ...               ...    ...   
2124477  c4a665596d4f67cecb7542c9fad407ee                 0  180.0   
2124488  c7b711619071c92bef604c7ad68380dd                 1  179.0   
2124558  dce1f234d6424aa61f8e7ce0baffd9af                 0  146.7   
2124597  e983f56659b43d6691446b013c0b9ef1                 0  149.2   
2124667  fe50ae64d08d4f8245aaabc55d1baf79                 1  143.2   

         total_sales                        st_city_id  \
3                6.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
13              12.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
34               0.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
70               4.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
85              10.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
...              ...                               ...   
2124477          3.0  885fe656777008c335ac96072a45be15   
2124488         10.0  885fe656777008c335ac96072a45be15   
2124558          3.0  885fe656777008c335ac96072a45be15   
2124597          4.0  885fe656777008c335ac96072a45be15   
2124667          5.0  885fe656777008c335ac96072a45be15   

                         st_division_code  st_type_format_id  st_type_loc_id  \
3        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
13       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
34       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
70       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
85       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
...                                   ...                ...             ...   
2124477  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124488  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124558  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124597  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124667  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   

         st_type_size_id  st_is_active  sales_total  \
3                      8             1     143898.0   
13                     8             1     143898.0   
34                     8             1     143898.0   
70                     8             1     143898.0   
85                     8             1     143898.0   
...                  ...           ...          ...   
2124477               12             1     100699.0   
2124488               12             1     100699.0   
2124558               12             1     100699.0   
2124597               12             1     100699.0   
2124667               12             1     100699.0   

                              pr_group_id                         pr_cat_id  \
3        6512bd43d9caa6e02c990b0a82652dca  e58cc5ca94270acaceed13bc82dfedf7   
13       c20

## Проверим признаки через PHIK

In [12]:
#check.drop('date', axis=1).phik_matrix()

In [13]:
display(check)

date                             st_id  \
3        2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
13       2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
34       2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
70       2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
85       2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
...             ...                               ...   
2124477  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124488  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124558  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124597  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124667  2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   

                                pr_sku_id  pr_sales_type_id  price  \
3        0094042bfeae507dc7f62acc8e5ed03a                 1  191.5   
13       050a8120e2a3b8751df8752fcc0323ef                 1  162.3   
34       0cf7b480289c0c4b07b6d3bd72fef0c9                 0    0.0   
70       194adbcaa46104c43d41ad4f23c18407                 1   53.0   
85       1cec786e41a51f75b5701511f995218b                 1   67.2   
...                                   ...               ...    ...   
2124477  c4a665596d4f67cecb7542c9fad407ee                 0  180.0   
2124488  c7b711619071c92bef604c7ad68380dd                 1  179.0   
2124558  dce1f234d6424aa61f8e7ce0baffd9af                 0  146.7   
2124597  e983f56659b43d6691446b013c0b9ef1                 0  149.2   
2124667  fe50ae64d08d4f8245aaabc55d1baf79                 1  143.2   

         total_sales                        st_city_id  \
3                6.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
13              12.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
34               0.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
70               4.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
85              10.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
...              ...                               ...   
2124477          3.0  885fe656777008c335ac96072a45be15   
2124488         10.0  885fe656777008c335ac96072a45be15   
2124558          3.0  885fe656777008c335ac96072a45be15   
2124597          4.0  885fe656777008c335ac96072a45be15   
2124667          5.0  885fe656777008c335ac96072a45be15   

                         st_division_code  st_type_format_id  st_type_loc_id  \
3        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
13       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
34       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
70       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
85       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
...                                   ...                ...             ...   
2124477  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124488  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124558  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124597  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124667  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   

         st_type_size_id  st_is_active  sales_total  \
3                      8             1     143898.0   
13                     8             1     143898.0   
34                     8             1     143898.0   
70                     8             1     143898.0   
85                     8             1     143898.0   
...                  ...           ...          ...   
2124477               12             1     100699.0   
2124488               12             1     100699.0   
2124558               12             1     100699.0   
2124597               12             1     100699.0   
2124667               12             1     100699.0   

                              pr_group_id                         pr_cat_id  \
3        6512bd43d9caa6e02c990b0a82652dca  e58cc5ca94270acaceed13bc82dfedf7   
13       c20

In [14]:
print("размер кластера", len(check)/352)

размер кластера 212.0


In [15]:
check.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 74624 entries, 3 to 2124667
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   date               74624 non-null  object 
 1   st_id              74624 non-null  object 
 2   pr_sku_id          74624 non-null  object 
 3   pr_sales_type_id   74624 non-null  int64  
 4   price              74624 non-null  float64
 5   total_sales        74624 non-null  float64
 6   st_city_id         74624 non-null  object 
 7   st_division_code   74624 non-null  object 
 8   st_type_format_id  74624 non-null  int64  
 9   st_type_loc_id     74624 non-null  int64  
 10  st_type_size_id    74624 non-null  int64  
 11  st_is_active       74624 non-null  int64  
 12  sales_total        74624 non-null  float64
 13  pr_group_id        74624 non-null  object 
 14  pr_cat_id          74624 non-null  object 
 15  pr_subcat_id       74624 non-null  object 
 16  pr_uom_id          7

In [16]:
check['date'] = pd.to_datetime(check['date'])

In [17]:
check['month'] = check.date.dt.month.astype(str)
check['day_of_week'] = check.date.dt.day_of_week.astype(str)
check['day_of_year'] = check.date.dt.day_of_year.astype(str)
check['day'] = check.date.dt.day.astype(str)
check['year_end'] = check.date.dt.is_year_end.astype(str)
check['year'] = check.date.dt.year.astype(str)
check

date                             st_id  \
3       2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
13      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
34      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
70      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
85      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
...            ...                               ...   
2124477 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124488 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124558 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124597 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124667 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   

                                pr_sku_id  pr_sales_type_id  price  \
3        0094042bfeae507dc7f62acc8e5ed03a                 1  191.5   
13       050a8120e2a3b8751df8752fcc0323ef                 1  162.3   
34       0cf7b480289c0c4b07b6d3bd72fef0c9                 0    0.0   
70       194adbcaa46104c43d41ad4f23c18407                 1   53.0   
85       1cec786e41a51f75b5701511f995218b                 1   67.2   
...                                   ...               ...    ...   
2124477  c4a665596d4f67cecb7542c9fad407ee                 0  180.0   
2124488  c7b711619071c92bef604c7ad68380dd                 1  179.0   
2124558  dce1f234d6424aa61f8e7ce0baffd9af                 0  146.7   
2124597  e983f56659b43d6691446b013c0b9ef1                 0  149.2   
2124667  fe50ae64d08d4f8245aaabc55d1baf79                 1  143.2   

         total_sales                        st_city_id  \
3                6.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
13              12.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
34               0.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
70               4.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
85              10.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
...              ...                               ...   
2124477          3.0  885fe656777008c335ac96072a45be15   
2124488         10.0  885fe656777008c335ac96072a45be15   
2124558          3.0  885fe656777008c335ac96072a45be15   
2124597          4.0  885fe656777008c335ac96072a45be15   
2124667          5.0  885fe656777008c335ac96072a45be15   

                         st_division_code  st_type_format_id  st_type_loc_id  \
3        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
13       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
34       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
70       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
85       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
...                                   ...                ...             ...   
2124477  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124488  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124558  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124597  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124667  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   

         ...                      pr_subcat_id  pr_uom_id  is_holiday cluster  \
3        ...  c93c26cd49ea08e6b8984351d4164575          1           0    11.0   
13       ...  f222c0d412b9a812f556597c95540276          1           0    11.0   
34       ...  45551f4c29881972ae5ed3666a5b1b8b          1           0    11.0   
70       ...  1507fc0175c1ff0cf4440105bc6f93e6         17           0    11.0   
85       ...  3dde95bfddfc5c387eb9e7b7a77dfa63          1           0    11.0   
...      ...                               ...        ...         ...     ...   
2124477  ...  45551f4c29881972ae5ed3666a5b1b8b          1           0    11.0   
2124488  ...  d29b5ce9c2883f0b7e90f79071a2ca82          1           0    11.0   
2124558  ...  c01c475ac8d447730ece97d846cc4f25          1           0    11.0   
2124597  ...  8b51a5273e5c6883ba

In [18]:
check.drop('date', axis=1).phik_matrix()

interval columns not set, guessing: ['pr_sales_type_id', 'price', 'total_sales', 'st_type_format_id', 'st_type_loc_id', 'st_type_size_id', 'st_is_active', 'sales_total', 'pr_uom_id', 'is_holiday', 'cluster']


st_id  pr_sku_id  pr_sales_type_id     price  \
st_id              1.000000   0.949884          0.135265  0.318105   
pr_sku_id          0.949884   1.000000          0.417155  0.880000   
pr_sales_type_id   0.135265   0.417155          1.000000  0.290346   
price              0.318105   0.880000          0.290346  1.000000   
total_sales        0.115955   0.422786          0.531558  0.168786   
st_city_id         1.000000   0.925928          0.117493  0.406332   
st_division_code   1.000000   0.933497          0.215077  0.310661   
st_type_format_id  1.000000   1.000000          0.008395  0.174375   
st_type_loc_id     1.000000   0.848118          0.031540  0.123868   
st_type_size_id    1.000000   0.980387          0.055934  0.238647   
sales_total        1.000000   0.933568          0.117500  0.424727   
pr_group_id        0.339301   1.000000          0.257537  0.658573   
pr_cat_id          0.584373   1.000000          0.250546  0.749531   
pr_subcat_id       0.821342   1.000000          0.347704  0.828378   
pr_uom_id          0.406085   1.000000          0.065961  0.761643   
is_holiday         0.000000   0.000000          0.000000  0.020882   
month              0.000000   0.000000          0.160930  0.081558   
day_of_week        0.000000   0.000000          0.136702  0.042261   
day_of_year        0.000000   0.000000          0.263555  0.079054   
day                0.000000   0.000000          0.022209  0.000000   
year_end           0.000000   0.000000          0.000000  0.000000   
year               0.000000   0.000000          0.077353  0.050385   

                   total_sales  st_city_id  st_division_code  \
st_id                 0.115955    1.000000          1.000000   
pr_sku_id             0.422786    0.925928          0.933497   
pr_sales_type_id      0.531558    0.117493          0.215077   
price                 0.168786    0.406332          0.310661   
total_sales           1.000000    0.159152          0.115548   
st_city_id            0.159152    1.000000          1.000000   
st_division_code      0.115548    1.000000          1.000000   
st_type_format_id     0.045116    0.186955          0.342302   
st_type_loc_id        0.056931    1.000000          0.943732   
st_type_size_id       0.148213    0.725773          0.611365   
sales_total           0.164221    0.995227          0.921027   
pr_group_id           0.190505    0.311593          0.314412   
pr_cat_id             0.219369    0.581299          0.577028   
pr_subcat_id          0.312866    0.798604          0.829427   
pr_uom_id             0.167755    0.332731          0.556074   
is_holiday            0.068318    0.000000          0.000000   
month                 0.165295    0.000000          0.000000   
day_of_week           0.244155    0.000000          0.000000   
day_of_year           0.420394    0.000000          0.000000   
day                   0.066042    0.000000          0.000000   
year_end              0.186229    0.000000          0.000000   
year                  0.062250    0.000000          0.000000   

                   st_type_format_id  st_type_loc_id  st_type_size_id  ...  \
st_id                       1.000000        1.000000         1.000000  ...   
pr_sku_id                   1.000000        0.848118         0.980387  ...   
pr_sales_type_id            0.008395        0.031540         0.055934  ...   
price                       0.174375        0.123868         0.238647  ...   
total_sales                 0.045116        0.056931         0.148213  ...   
st_city_id                  0.186955        1.000000         0.725773  ...   
st_division_code            0.342302        0.943732         0.611365  ...   
st_type_format_id           1.000000        0.104502         1.000000  ...   
st_type_loc_id              0.104502        1.000000         0.340894  ...   
st_type_size_id             1.000000        0.340894         1.000000  ...   
sales_total                      NaN        1.000000         1.000000  ...   
pr

Если бы я удалял колонки, то в первую очередь бы удалил : day, month, year, pr_uom_id, st_division_code, st_type_loc_id.

In [19]:
check = check.drop(['year', 'day', 'sales_total', 'cluster', 'st_is_active'], axis = 1)

In [20]:
check

date                             st_id  \
3       2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
13      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
34      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
70      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
85      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
...            ...                               ...   
2124477 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124488 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124558 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124597 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124667 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   

                                pr_sku_id  pr_sales_type_id  price  \
3        0094042bfeae507dc7f62acc8e5ed03a                 1  191.5   
13       050a8120e2a3b8751df8752fcc0323ef                 1  162.3   
34       0cf7b480289c0c4b07b6d3bd72fef0c9                 0    0.0   
70       194adbcaa46104c43d41ad4f23c18407                 1   53.0   
85       1cec786e41a51f75b5701511f995218b                 1   67.2   
...                                   ...               ...    ...   
2124477  c4a665596d4f67cecb7542c9fad407ee                 0  180.0   
2124488  c7b711619071c92bef604c7ad68380dd                 1  179.0   
2124558  dce1f234d6424aa61f8e7ce0baffd9af                 0  146.7   
2124597  e983f56659b43d6691446b013c0b9ef1                 0  149.2   
2124667  fe50ae64d08d4f8245aaabc55d1baf79                 1  143.2   

         total_sales                        st_city_id  \
3                6.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
13              12.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
34               0.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
70               4.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
85              10.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
...              ...                               ...   
2124477          3.0  885fe656777008c335ac96072a45be15   
2124488         10.0  885fe656777008c335ac96072a45be15   
2124558          3.0  885fe656777008c335ac96072a45be15   
2124597          4.0  885fe656777008c335ac96072a45be15   
2124667          5.0  885fe656777008c335ac96072a45be15   

                         st_division_code  st_type_format_id  st_type_loc_id  \
3        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
13       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
34       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
70       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
85       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
...                                   ...                ...             ...   
2124477  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124488  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124558  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124597  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124667  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   

         st_type_size_id                       pr_group_id  \
3                      8  6512bd43d9caa6e02c990b0a82652dca   
13                     8  c20ad4d76fe97759aa27a0c99bff6710   
34                     8  c20ad4d76fe97759aa27a0c99bff6710   
70                     8  c74d97b01eae257e44aa9d5bade97baf   
85                     8  c51ce410c124a10e0db5e4b97fc2af39   
...                  ...                               ...   
2124477               12  c20ad4d76fe97759aa27a0c99bff6710   
2124488               12  aab3238922bcc25a6f606eb525ffdc56   
2124558               12  aab3238922bcc25a6f606eb525ffdc56   
2124597               12  aab3238922bcc25a6f606eb525ffdc56   
2124667               12  aab3238922bcc25a6f606eb525ffdc56   

                                pr_cat_id                      pr_subcat_id  \
3        e58cc5ca9

делим на выборки, чтобы не было утечки целевого признака

In [21]:
train_class = check[check.date<='2023-07-04']
test_class = check[check.date>'2023-07-04']

## Лаги и средние скользящие

In [22]:
#check = check.sort_values(by=['date', 'st_id', 'pr_sku_id'])

In [23]:
for n in range(7, 71, 7):
    check[f'lag{n}'] = check.total_sales\
        .shift(check.date.value_counts().values[0]*n)

In [24]:
for m in range(7, 36, 7):
    check[f'rolling_mean_{m}'] = check.total_sales.rolling(check.date.value_counts().values[0]*m).mean()

In [25]:
check

date                             st_id  \
3       2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
13      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
34      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
70      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
85      2022-08-01  16a5cdae362b8d27a1d8f8c7b78b4330   
...            ...                               ...   
2124477 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124488 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124558 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124597 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   
2124667 2023-07-18  fa7cdfad1a5aaf8370ebeda47a1ff1c3   

                                pr_sku_id  pr_sales_type_id  price  \
3        0094042bfeae507dc7f62acc8e5ed03a                 1  191.5   
13       050a8120e2a3b8751df8752fcc0323ef                 1  162.3   
34       0cf7b480289c0c4b07b6d3bd72fef0c9                 0    0.0   
70       194adbcaa46104c43d41ad4f23c18407                 1   53.0   
85       1cec786e41a51f75b5701511f995218b                 1   67.2   
...                                   ...               ...    ...   
2124477  c4a665596d4f67cecb7542c9fad407ee                 0  180.0   
2124488  c7b711619071c92bef604c7ad68380dd                 1  179.0   
2124558  dce1f234d6424aa61f8e7ce0baffd9af                 0  146.7   
2124597  e983f56659b43d6691446b013c0b9ef1                 0  149.2   
2124667  fe50ae64d08d4f8245aaabc55d1baf79                 1  143.2   

         total_sales                        st_city_id  \
3                6.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
13              12.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
34               0.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
70               4.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
85              10.0  c1f75cc0f7fe269dd0fd9bd5e24f9586   
...              ...                               ...   
2124477          3.0  885fe656777008c335ac96072a45be15   
2124488         10.0  885fe656777008c335ac96072a45be15   
2124558          3.0  885fe656777008c335ac96072a45be15   
2124597          4.0  885fe656777008c335ac96072a45be15   
2124667          5.0  885fe656777008c335ac96072a45be15   

                         st_division_code  st_type_format_id  st_type_loc_id  \
3        296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
13       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
34       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
70       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
85       296bd0cc6e735f9d7488ebc8fbc19130                  1               2   
...                                   ...                ...             ...   
2124477  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124488  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124558  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124597  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   
2124667  296bd0cc6e735f9d7488ebc8fbc19130                  1               1   

         ...  lag42 lag49 lag56 lag63  lag70  rolling_mean_7 rolling_mean_14  \
3        ...    NaN   NaN   NaN   NaN    NaN             NaN             NaN   
13       ...    NaN   NaN   NaN   NaN    NaN             NaN             NaN   
34       ...    NaN   NaN   NaN   NaN    NaN             NaN             NaN   
70       ...    NaN   NaN   NaN   NaN    NaN             NaN             NaN   
85       ...    NaN   NaN   NaN   NaN    NaN             NaN             NaN   
...      ...    ...   ...   ...   ...    ...             ...             ...   
2124477  ...    5.0  10.0   5.0   4.0   22.0       11.839454       11.916400   
2124488  ...   11.0  14.0   7.0   6.0   17.0       11.836085       11.913705   
2124558  ...    2.0   4.0   4.0   2.0    9.0       11.834063       11.911346   
2124597  ...    5.0  11.0   2.0   4.0   15

## Кодирование категориальных признаков

In [26]:
cat_col = ['st_id', 'pr_sku_id', 'pr_sales_type_id', 'pr_group_id', 'pr_cat_id', 'pr_subcat_id',
       'pr_uom_id', 'st_city_id', 'st_division_code', 'st_type_format_id',
       'st_type_loc_id', 'st_type_size_id', 'is_holiday', 'month', 'day_of_week', 'day_of_year', 'year_end']
num_col = ['price',
       'total_sales']
nums = ['lag7', 'lag14', 'lag21', 'lag28', 'lag35', 'lag42', 'lag49', 'lag56', 'lag63',
    'lag70', 'rolling_mean_7', 'rolling_mean_14', 'rolling_mean_21', 'rolling_mean_28', 'rolling_mean_35']

In [27]:
enc = TargetEncoder(target_type='continuous', smooth='auto')

In [28]:
enc.fit(check[cat_col], check['total_sales'])
enc_cols = enc.get_feature_names_out(check[cat_col].columns)

In [29]:
check_enc = pd.DataFrame(enc.transform(check[cat_col]), columns=enc_cols,
 index=check[cat_col].index)

In [30]:
check_enc

st_id  pr_sku_id  pr_sales_type_id  pr_group_id  pr_cat_id  \
3        9.495453   7.630256         12.169702     8.091308   7.879414   
13       9.495453   6.204158         12.169702     8.242103   8.242103   
34       9.495453   8.898847          4.538560     8.242103   8.242103   
70       9.495453   9.272398         12.169702     9.440768   8.733654   
85       9.495453  10.666302         12.169702    10.252849  10.673022   
...           ...        ...               ...          ...        ...   
2124477  9.125596  10.199314          4.538560     8.242103   8.242103   
2124488  9.125596  12.220641         12.169702     9.455236   9.591152   
2124558  9.125596   8.532494          4.538560     9.455236   8.785339   
2124597  9.125596   7.118653          4.538560     9.455236   6.671158   
2124667  9.125596   8.473885         12.169702     9.455236   9.591152   

         pr_subcat_id  pr_uom_id  st_city_id  st_division_code  \
3            7.951768   8.731174    9.495453          9.380684   
13           6.096275   8.731174    9.495453          9.380684   
34           8.456692   8.731174    9.495453          9.380684   
70           9.272398   9.385851    9.495453          9.380684   
85          10.666302   8.731174    9.495453          9.380684   
...               ...        ...         ...               ...   
2124477      8.456692   8.731174    9.125596          9.380684   
2124488     12.173662   8.731174    9.125596          9.380684   
2124558      8.532494   8.731174    9.125596          9.380684   
2124597      6.671158   8.731174    9.125596          9.380684   
2124667      9.762588   8.731174    9.125596          9.380684   

         st_type_format_id  st_type_loc_id  st_type_size_id  is_holiday  \
3                 8.983594        9.495453         9.311345    8.866802   
13                8.983594        9.495453         9.311345    8.866802   
34                8.983594        9.495453         9.311345    8.866802   
70                8.983594        9.495453         9.311345    8.866802   
85                8.983594        9.495453         9.311345    8.866802   
...                    ...             ...              ...         ...   
2124477           8.983594        8.823674         8.718543    8.866802   
2124488           8.983594        8.823674         8.718543    8.866802   
2124558           8.983594        8.823674         8.718543    8.866802   
2124597           8.983594        8.823674         8.718543    8.866802   
2124667           8.983594        8.823674         8.718543    8.866802   

             month  day_of_week  day_of_year  year_end  
3        10.217959     6.571465     7.619229  8.927994  
13       10.217959     6.571465     7.619229  8.927994  
34       10.217959     6.571465     7.619229  8.927994  
70       10.217959     6.571465     7.619229  8.927994  
85       10.217959     6.571465     7.619229  8.927994  
...            ...          ...          ...       ...  
2124477  11.871586     6.725168     8.508813  8.927994  
2124488  11.871586     6.725168     8.508813  8.927994  
2124558  11.871586     6.725168     8.508813  8.927994  
2124597  11.871586     6.725168     8.508813  8.927994  
2124667  11.871586     6.725168     8.508813  8.927994  

[74624 rows x 17 columns]

In [31]:
check_enc[nums] = check[nums]

In [32]:
check_enc[num_col] = check[num_col]

In [33]:
check_enc['date'] = check['date']

In [34]:
check_enc

st_id  pr_sku_id  pr_sales_type_id  pr_group_id  pr_cat_id  \
3        9.495453   7.630256         12.169702     8.091308   7.879414   
13       9.495453   6.204158         12.169702     8.242103   8.242103   
34       9.495453   8.898847          4.538560     8.242103   8.242103   
70       9.495453   9.272398         12.169702     9.440768   8.733654   
85       9.495453  10.666302         12.169702    10.252849  10.673022   
...           ...        ...               ...          ...        ...   
2124477  9.125596  10.199314          4.538560     8.242103   8.242103   
2124488  9.125596  12.220641         12.169702     9.455236   9.591152   
2124558  9.125596   8.532494          4.538560     9.455236   8.785339   
2124597  9.125596   7.118653          4.538560     9.455236   6.671158   
2124667  9.125596   8.473885         12.169702     9.455236   9.591152   

         pr_subcat_id  pr_uom_id  st_city_id  st_division_code  \
3            7.951768   8.731174    9.495453          9.380684   
13           6.096275   8.731174    9.495453          9.380684   
34           8.456692   8.731174    9.495453          9.380684   
70           9.272398   9.385851    9.495453          9.380684   
85          10.666302   8.731174    9.495453          9.380684   
...               ...        ...         ...               ...   
2124477      8.456692   8.731174    9.125596          9.380684   
2124488     12.173662   8.731174    9.125596          9.380684   
2124558      8.532494   8.731174    9.125596          9.380684   
2124597      6.671158   8.731174    9.125596          9.380684   
2124667      9.762588   8.731174    9.125596          9.380684   

         st_type_format_id  ...  lag63  lag70  rolling_mean_7  \
3                 8.983594  ...    NaN    NaN             NaN   
13                8.983594  ...    NaN    NaN             NaN   
34                8.983594  ...    NaN    NaN             NaN   
70                8.983594  ...    NaN    NaN             NaN   
85                8.983594  ...    NaN    NaN             NaN   
...                    ...  ...    ...    ...             ...   
2124477           8.983594  ...    4.0   22.0       11.839454   
2124488           8.983594  ...    6.0   17.0       11.836085   
2124558           8.983594  ...    2.0    9.0       11.834063   
2124597           8.983594  ...    4.0   15.0       11.833389   
2124667           8.983594  ...    7.0   12.0       11.831368   

         rolling_mean_14  rolling_mean_21  rolling_mean_28  rolling_mean_35  \
3                    NaN              NaN              NaN              NaN   
13                   NaN              NaN              NaN              NaN   
34                   NaN              NaN              NaN              NaN   
70                   NaN              NaN              NaN              NaN   
85                   NaN              NaN              NaN              NaN   
...                  ...              ...              ...              ...   
2124477        11.916400        11.935170        11.751937        11.731014   
2124488        11.913705        11.934046        11.752274        11.731149   
2124558        11.911346        11.932699        11.751600        11.731149   
2124597        11.910672        11.932025        11.751600        11.731149   
2124667        11.910335        11.931351        11.751263        11.731149   

         price  total_sales       date  
3        191.5          6.0 2022-08-01  
13       162.3         12.0 2022-08-01  
34         0.0          0.0 2022-08-01  
70        53.0          4.0 2022-08-01  
85        67.2         10.0 2022-08-01  
...        ...          ...        ...  
2124477  180.0          3.0 2023-07-18  
2124488  179.0         10.0 2023-07-18  
2124558  146.7          3.0 2023-07-18  
2124597  149.2          4.0 2023-07-18  
2124667  143.2          5.0 2023-07-18  

[74624 rows x 35 columns]

### перед разбивкой надо удалить нули

In [35]:
check_enc.dropna(inplace=True)

## Разбивка на выборки

In [36]:
check_enc.index = check_enc.date
check_enc.index.is_monotonic_increasing

True

In [37]:
#check_enc = pd.to_numeric(check_enc['date'], errors='coerce')

In [38]:
train_data = check_enc[check_enc.date<='2023-07-04']
test_data = check_enc[check_enc.date>'2023-07-04']

In [39]:
test_data

st_id  pr_sku_id  pr_sales_type_id  pr_group_id  pr_cat_id  \
date                                                                        
2023-07-05  9.495453   7.630256          4.538560     8.091308   7.879414   
2023-07-05  9.495453   6.204158         12.169702     8.242103   8.242103   
2023-07-05  9.495453   8.898847          4.538560     8.242103   8.242103   
2023-07-05  9.495453   9.272398         12.169702     9.440768   8.733654   
2023-07-05  9.495453  10.666302         12.169702    10.252849  10.673022   
...              ...        ...               ...          ...        ...   
2023-07-18  9.125596  10.199314          4.538560     8.242103   8.242103   
2023-07-18  9.125596  12.220641         12.169702     9.455236   9.591152   
2023-07-18  9.125596   8.532494          4.538560     9.455236   8.785339   
2023-07-18  9.125596   7.118653          4.538560     9.455236   6.671158   
2023-07-18  9.125596   8.473885         12.169702     9.455236   9.591152   

            pr_subcat_id  pr_uom_id  st_city_id  st_division_code  \
date                                                                
2023-07-05      7.951768   8.731174    9.495453          9.380684   
2023-07-05      6.096275   8.731174    9.495453          9.380684   
2023-07-05      8.456692   8.731174    9.495453          9.380684   
2023-07-05      9.272398   9.385851    9.495453          9.380684   
2023-07-05     10.666302   8.731174    9.495453          9.380684   
...                  ...        ...         ...               ...   
2023-07-18      8.456692   8.731174    9.125596          9.380684   
2023-07-18     12.173662   8.731174    9.125596          9.380684   
2023-07-18      8.532494   8.731174    9.125596          9.380684   
2023-07-18      6.671158   8.731174    9.125596          9.380684   
2023-07-18      9.762588   8.731174    9.125596          9.380684   

            st_type_format_id  ...  lag63  lag70  rolling_mean_7  \
date                           ...                                 
2023-07-05           8.983594  ...    6.0    3.0       11.974057   
2023-07-05           8.983594  ...    0.0    0.0       11.981469   
2023-07-05           8.983594  ...    7.0   11.0       11.983491   
2023-07-05           8.983594  ...    9.0   10.0       11.978100   
2023-07-05           8.983594  ...   10.0   19.0       11.977426   
...                       ...  ...    ...    ...             ...   
2023-07-18           8.983594  ...    4.0   22.0       11.839454   
2023-07-18           8.983594  ...    6.0   17.0       11.836085   
2023-07-18           8.983594  ...    2.0    9.0       11.834063   
2023-07-18           8.983594  ...    4.0   15.0       11.833389   
2023-07-18           8.983594  ...    7.0   12.0       11.831368   

            rolling_mean_14  rolling_mean_21  rolling_mean_28  \
date                                                            
2023-07-05        11.589833        11.611467        11.904881   
2023-07-05        11.593876        11.614611        11.906566   
2023-07-05        11.590844        11.613488        11.907071   
2023-07-05        11.588822        11.611916        11.905892   
2023-07-05        11.587138        11.612814        11.904207   
...                     ...              ...              ...   
2023-07-18        11.916400        11.935170        11.751937   
2023-07-18        11.913705        11.934046        11.752274   
2023-07-18        11.911346        11.932699        11.751600   
2023-07-18        11.910672        11.932025        11.751600   
2023-07-18        11.910335        11.931351        11.751263   

            rolling_mean_35  price  total_sales       date  
date                                                        
2023-07-05        11.788022  204.2          5.0 2023-07-05  
2023-07-05        11.789370  164.6         14.0 2023-07-05  
2023-07-05        11.789774  126.3          3.0 2023-07-05  
2023-07-05        11.788966   42.0         10.0 2023-07-05  
2023-07-05        

In [40]:
x_train = train_data.drop('total_sales', axis=1)
y_train = train_data['total_sales']

In [41]:
#print(x_train.dtypes)
#print(y_train.dtype)

In [42]:
x_test = test_data.drop('total_sales', axis=1)
y_test = test_data['total_sales']

In [43]:
def wape(y_true, y_pred):
    return np.sum(np.abs(y_true-y_pred))/np.sum(np.abs(y_true))

In [44]:
score = make_scorer(wape, greater_is_better=False)

In [45]:
#x_train = x_train.drop('date', axis = 1)
x_train['date'] = pd.to_numeric(x_train['date'], errors='coerce')

In [46]:
tscv = TimeSeriesSplit(n_splits=6)
lr = LinearRegression()
parametrs = {}
lr_grid = GridSearchCV(lr, parametrs, scoring=score, cv=tscv)
lr_grid.fit(x_train, y_train)

GridSearchCV(cv=TimeSeriesSplit(gap=0, max_train_size=None, n_splits=6, test_size=None),
             estimator=LinearRegression(), param_grid={},
             scoring=make_scorer(wape, greater_is_better=False))

In [47]:
x_test['date'] = pd.to_numeric(x_test['date'], errors='coerce')

In [48]:
lr_preds = lr_grid.best_estimator_.predict(x_test)

In [49]:
wape(y_test, lr_preds)

0.4475507065753344

# Наблюдения

кластер 7 wape 0.499 размер 57

кластер 34 wape = 0.31495 размер 28

кластер 11 wape 0.4475 размер 212

кластер 9 wape 0.706 размер 5564

кластер 0 wape 0.86 размер 66

кластер 4 wape 0.3119 размер 21

кластер 5 wape 0.36285 размер 12

кластер 3 wape 0.25296 размер 8

кластер 1 wape 1.3888 размер 6

кластер 13 wape 0.16729 размер 5

кластер 18 wape 0.5 размер 4

кластер 17 wape 0.18 размер 4

кластер 17 wape 0.28 размер 3

кластер 2 wape 0.1185 размер 3

кластер 14 wape 0.2088 размер 2

кластер 12 wape 0.1167 размер 2

кластер 21 wape 1.742824666734505e-05 размер 1

кластер 35 wape 0.01689 размер 1

кластер 20 wape 0.000542 размер 1

кластер 38 wape 0.000546 размер 1

## ПРоблема нулей! 

есть целые маленькие кластеры нулей, так и просто сегменты, где все значения нули.... 

Да и вообще, мне кажется, очень много лишнего. 

GIGO 